refactor: cleanup/followup of Remove -deprecatedrpc=addresses flag

mjdietzx opened the pull request:
This is a follow-up on top of #22650. Based on review there, we thought it would be an improvement to merge `ScriptPubKeyToUniv` and `ScriptToUniv` into a single function, but this didn't seem to belong in that PR. There are a few other minor cleanups here as well focused on style and making some fuzz tests run faster

mjdietzx commented:
Marking this as a draft for now. I will be moving some of the refactoring/cleanup that was originally included in #22650 to this PR. So once that is done and I've rebased this PR, I will reopen it

MarcoFalke commented on src/test/fuzz/script.cpp:
```diff
@@ -56,46 +56,8 @@ FUZZ_TARGET_INIT(script, initialize_script)
-        }
-    } else {
-        assert(required_ret >= 1 && required_ret <= 16);
-        assert((unsigned long)required_ret == addresses.size());
```
This assertion fails:

https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=39152

It might be good to merge this pull request or apply the changes to this fuzz test early.

mjdietzx committed:
refactor: merge ScriptPubKeyToUniv & ScriptToUniv into one function

mjdietzx committed:
refactor: named args whenever ScriptToUniv or TxToUniv invoked

mjdietzx committed:
refactor: de-duplicate common help text for parsed destination address

mjdietzx commented:
Now that #22650 is merged, I rebased and finished this. This PR follows up with some cleanup, and addresses all the comments in the first PR that were left for now.

Just so we have them at-hand, I took all the original comments from #22650's review, and am putting them here.

> @jonatack: maybe I'm missing something but perhaps merge ScriptPubKeyToUniv and ScriptToUniv into a single function

Done b9ad1d9

---

> @jonatack: Named args could be nice wherever this function is invoked

and

> @MarcoFalke: If you decide to do it here, at least use the proper clang-tidy format, so that it can be checked by the clang compiler

Done e9f3f1b

---

> @MarcoFalke: as a follow-up it could make sense to de-duplicate the common help text

Done fe71ac8

---

> @MarcoFalke: I think you can drop one of these statements to make the fuzz test run ~twice as fast. An alternative (to test both code paths) would be to deserialize a bool from the fuzzing input and use uint256{bool_val}

and

> @meshcollider: This should really be replaced with include_address=true/false checks on ScriptPubKeyToUniv(script, o4, fIncludeHex, include_address)

Done e749168

MarcoFalke commented on src/test/fuzz/transaction.cpp:
```diff
@@ -103,6 +105,5 @@ FUZZ_TARGET_INIT(transaction, initialize_transaction)
     UniValue u(UniValue::VOBJ);
-    TxToUniv(tx, /* hashBlock */ uint256::ZERO, u);
-    TxToUniv(tx, /* hashBlock */ uint256::ONE, u);
+    TxToUniv(tx, /* hashBlock= */ uint256{fuzzed_data_provider.ConsumeBool()}, /* entry= */ u, fuzzed_data_provider.ConsumeBool());
```
not a fan of re-using the same buffer, thus nVersion for this bool.

`try { ds >> bool{}; }` might be nicer

mjdietzx replied:
Makes sense, done 08543bf. I may have butchered this, so if you want to see any other changes here just lmk. Happy to fix up any nits bc I have a feeling there is a cleaner way to do this that I'm just not aware of

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1762,7 +1762,7 @@ static RPCHelpMan gettransaction()
     if (verbose) {
         UniValue decoded(UniValue::VOBJ);
-        TxToUniv(*wtx.tx, uint256(), decoded, false);
+        TxToUniv(*wtx.tx, /* hashBlock= */ uint256(), /* entry= */ decoded, /* include_hex= */ false);
```
nit: Maybe before using named args, fix the argument name first?

```suggestion
        TxToUniv(*wtx.tx, /*block_hash=*/uint256{}, /*entry=*/decoded, /*include_hex=*/false);
```

mjdietzx replied:
Added a commit, done here: be3651a. I had wanted to do this but seemed like `hashBlock` was used pretty regularly so I thought there might be some convention I wasn't aware of - I guess not so I'm happy to be able to do this

mjdietzx committed:
refactor: prefer snake case, TxToUniv arg hashBlock renamed block_hash

MarcoFalke commented on src/test/fuzz/transaction.cpp:
```diff
@@ -102,7 +102,14 @@ FUZZ_TARGET_INIT(transaction, initialize_transaction)
+        ds >> block_hash;
+        ds >> include_hex;
+    } catch (const std::ios_base::failure&) {
+        return;
```
If you initialize the bools to `false`, you can remove the early return (This is also what FuzzedDataProvider would do)

mjdietzx replied:
That's a good idea. Ok, one sec

mjdietzx committed:
fuzz: speedup ScriptToUniv and TxToUniv tests by dropping redundant call

laanwj commented on src/core_write.cpp:
```diff
@@ -142,22 +142,19 @@ std::string EncodeHexTx(const CTransaction& tx, const int serializeFlags)
-    if (include_hex) out.pushKV("hex", HexStr(scriptPubKey));
+    out.pushKV("asm", ScriptToAsmStr(script));
+    if (include_hex) {
+      out.pushKV("hex", HexStr(script));
```
indentation nit

mjdietzx replied:
Nice catch, done.

laanwj commented on src/core_io.h:
```diff
@@ -53,8 +53,7 @@ UniValue ValueFromAmount(const CAmount amount);
-void ScriptPubKeyToUniv(const CScript& scriptPubKey, UniValue& out, bool include_hex, bool include_address = true);
-void ScriptToUniv(const CScript& script, UniValue& out);
-void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry, bool include_hex = true, int serialize_flags = 0, const CTxUndo* txundo = nullptr, TxVerbosity verbosity = TxVerbosity::SHOW_DETAILS);
+void ScriptToUniv(const CScript& script, UniValue& out, bool include_hex = true, bool include_address = false);
```
I wonder why we don't return the UniValue (or `optional<UniValue>` in case it can fail) instead of having it as second parameter. It would seem better API design to me.

mjdietzx replied:
I think this is a great suggestion. And at first glance it seems to be straightforward enough to include in this PR. I'll follow up shortly

mjdietzx replied:
I am leaning towards not including this change in this PR. I do think it is a great opportunity to cleanup and improve the code quality, so would like to follow up with another PR doing this. This pattern is also used all over the place, not just `ScriptToUniv`, but `TxToUniv`, `entryToJSON`, and probably more. Anyways I want to spend some more time on this and don't want to end up doing too much in this PR

mjdietzx replied:
Can you take a quick look at the last two commits in this draft PR #23507?

a) is this along the lines of what you are thinking?
b) do you think it is worthwhile for me to do this across the codebase in #23507, eg `entryToJSON`, `TxToJSON`, etc... can be improved in the same way

Feel free to answer those in the linked PR and not here

mjdietzx commented:
Addressed indentation nit and rebased to master

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -1274,7 +1274,7 @@ static RPCHelpMan gettxout()
                     {RPCResult::Type::STR_HEX, "hex", ""},
                     {RPCResult::Type::STR, "type", "The type, eg pubkeyhash"},
-                    {RPCResult::Type::STR, "address", /* optional */ true, "The Bitcoin address (only if a well-defined address exists)"},
+                    {RPCResult::Type::STR, "address", /* optional */ true, DESTINATION_ADDRESS},
```
it might be better to de-duplicate the whole asm/hex/type/address blob. Though, maybe create a separate pull for doc cleanups vs code refactorings?

MarcoFalke replied:
See for example e7507f333b or f8911de619d41a3cc21771d62ab077713f1405c5 on how to do this.

mjdietzx replied:
I think this is a good suggestion. I will do this as a followup as you suggest. So I will get rid of 63a657ccff959f7d86f423607f69aa9f7507dbdb when I rebase and have a doc cleanup followup with these suggestions

fanquake commented:
I've rebased some of the commits here in #24673. Haven't done 63a657ccff959f7d86f423607f69aa9f7507dbdb as we might be able to follow up differently there, and do some de-duplication, see https://github.com/bitcoin/bitcoin/pull/22924#discussion_r760034440.

mjdietzx commented:
Thanks @fanquake - it seems like the best path forward is closing this PR in favor of https://github.com/bitcoin/bitcoin/pull/24673? If so I'll close this and test/review the new PR
